# HG changeset patch
# User Jason Kim <jasonwkim@google.com>
# Date 1301356919 25200
# Branch pnacl-sfi
# Node ID bb23f96ef85abf5287e4aad9ed511efd7300705f
# Parent 441feeabe2c21cca8eb92cb2ee3b61d39e339b82
1. Add support for new ARM triple arm-nacl-gnueabi
2. Invert SFI switches. Now they DISABLE specific SFI specific features
3.

 From llvm-pnacl-0018-240-bb23f96ef85abf5287e4aad9ed511efd7300705f.patch

 From .hg/patches/llvm-pnacl-0018-240-bb23f96ef85abf5287e4aad9ed511efd7300705f.patch.stripped

diff -r 441feeabe2c2 lib/Support/Triple.cpp
--- a/lib/Support/Triple.cpp	Mon Mar 28 17:01:56 2011 -0700
+++ b/lib/Support/Triple.cpp	Tue Jun 28 10:56:00 2011 -0700
@@ -317,8 +317,8 @@
     return FreeBSD;
   else if (OSName.startswith("linux"))
     // TODO(pdox): Fix this when we stop using target linux
-    //return Linux;
-    return NativeClient; // @LOCALMOD
+    return Linux;
+    //return NativeClient; // @LOCALMOD
   else if (OSName.startswith("lv2"))
     return Lv2;
   else if (OSName.startswith("mingw32"))
diff -r 441feeabe2c2 lib/Target/ARM/ARMAsmPrinter.cpp
--- a/lib/Target/ARM/ARMAsmPrinter.cpp	Mon Mar 28 17:01:56 2011 -0700
+++ b/lib/Target/ARM/ARMAsmPrinter.cpp	Tue Jun 28 10:56:00 2011 -0700
@@ -58,8 +58,8 @@
 
 // @LOCALMOD-START
 namespace llvm {
-  extern cl::opt<bool> FlagSfiBranch;
-  extern cl::opt<bool> FlagSfiData;
+  extern cl::opt<bool> FlagSfiDisableBranch;
+  extern cl::opt<bool> FlagSfiDisableData;
 }
 // @LOCALMOD-END
 
@@ -265,7 +265,7 @@
   MCP = MF.getConstantPool();
 
   // @LOCALMOD-START
-  if (FlagSfiBranch) {
+  if (Subtarget->isTargetNaCl() && (!FlagSfiDisableBranch)) {
     NaclAlignAllJumpTargetsAndConstantPools(MF);
   }
   // @LOCALMOD-END
@@ -1407,7 +1407,7 @@
         OutStreamer.EmitRawText(StringRef("sfi_nop_if_at_bundle_end\n"));
       }
 
-      if (FlagSfiData) {
+      if (!FlagSfiDisableData) {
         SmallString<128> Str;
         raw_svector_ostream OS(Str);
         OS << "sfi_illegal_if_at_bundle_begining  @ ========== SFI (" << 
diff -r 441feeabe2c2 lib/Target/ARM/ARMAsmPrinter.h
--- a/lib/Target/ARM/ARMAsmPrinter.h	Mon Mar 28 17:01:56 2011 -0700
+++ b/lib/Target/ARM/ARMAsmPrinter.h	Tue Jun 28 10:56:00 2011 -0700
@@ -73,7 +73,8 @@
   // For the sfi case we do not use the custom logic and fall back
   // to the default implementation.
   virtual void EmitConstantPool() {
-    if (FlagSfiDisableCP) AsmPrinter::EmitConstantPool();
+    if (Subtarget->isTargetNaCl() && (!FlagSfiEnableCP))
+      AsmPrinter::EmitConstantPool();
   }
   // @LOCALMOD-END
 
diff -r 441feeabe2c2 lib/Target/ARM/ARMBaseRegisterInfo.cpp
--- a/lib/Target/ARM/ARMBaseRegisterInfo.cpp	Mon Mar 28 17:01:56 2011 -0700
+++ b/lib/Target/ARM/ARMBaseRegisterInfo.cpp	Tue Jun 28 10:56:00 2011 -0700
@@ -852,7 +852,7 @@
   // Sadly, the ARM backend is not very consistent about using this
   // pseudo instr. and hence checking this is not sufficient.
   // But, it should help detect some regressions early.
-  assert(!FlagSfiDisableCP && "unexpected call to emitLoadConstPool");
+  assert ((!STI.isTargetNaCl()) && "unexpected call to emitLoadConstPool");
   // @LOCALMOD-END
   MachineFunction &MF = *MBB.getParent();
   MachineConstantPool *ConstantPool = MF.getConstantPool();
diff -r 441feeabe2c2 lib/Target/ARM/ARMConstantIslandPass.cpp
--- a/lib/Target/ARM/ARMConstantIslandPass.cpp	Mon Mar 28 17:01:56 2011 -0700
+++ b/lib/Target/ARM/ARMConstantIslandPass.cpp	Tue Jun 28 10:56:00 2011 -0700
@@ -319,7 +319,6 @@
 }
 
 bool ARMConstantIslands::runOnMachineFunction(MachineFunction &MF) {
-  if (FlagSfiDisableCP) return false;   // @LOCALMOD
   MachineConstantPool &MCP = *MF.getConstantPool();
 
   TII = (const ARMInstrInfo*)MF.getTarget().getInstrInfo();
@@ -327,6 +326,9 @@
   AFI = MF.getInfo<ARMFunctionInfo>();
   STI = &MF.getTarget().getSubtarget<ARMSubtarget>();
 
+  if (STI->isTargetNaCl() && (!FlagSfiEnableCP)) // @LOCALMOD
+    return false;   // @LOCALMOD
+  
   isThumb = AFI->isThumbFunction();
   isThumb1 = AFI->isThumb1OnlyFunction();
   isThumb2 = AFI->isThumb2Function();
@@ -624,7 +626,7 @@
 }
 
 static void UpdateJumpTargetAlignment(MachineFunction &MF) {
-  if (!FlagSfiBranch) return;
+  if (FlagSfiDisableBranch) return;
 
   // JUMP TABLE TARGETS
   MachineJumpTableInfo *jt_info = MF.getJumpTableInfo();
diff -r 441feeabe2c2 lib/Target/ARM/ARMExpandPseudoInsts.cpp
--- a/lib/Target/ARM/ARMExpandPseudoInsts.cpp	Mon Mar 28 17:01:56 2011 -0700
+++ b/lib/Target/ARM/ARMExpandPseudoInsts.cpp	Tue Jun 28 10:56:00 2011 -0700
@@ -730,8 +730,8 @@
   // We need to know if it matters that references are pc-relative
   // (e.g., to be PIC).
   // See the comments on MOVi16PIC / MOVTi16PIC for more details.
-  const bool ShouldUseMOV16PIC = FlagSfiDisableCP && IsRelocPIC &&
-      (MO.isCPI() || MO.isJTI() || MO.isGlobal()); // TODO check this list.
+  const bool ShouldUseMOV16PIC = (!FlagSfiEnableCP) && IsRelocPIC &&
+    (STI->isTargetNaCl() ||  MO.isCPI() || MO.isJTI() || MO.isGlobal()); // TODO check this list.
   if (ShouldUseMOV16PIC) {
     if (isThumb2)
       llvm_unreachable("FIXME: add PIC versions of t2MOVi16");
diff -r 441feeabe2c2 lib/Target/ARM/ARMFastISel.cpp
--- a/lib/Target/ARM/ARMFastISel.cpp	Mon Mar 28 17:01:56 2011 -0700
+++ b/lib/Target/ARM/ARMFastISel.cpp	Tue Jun 28 10:56:00 2011 -0700
@@ -596,7 +596,8 @@
   // @LOCALMOD-START
   // In the sfi case we do not want to use the ARM custom cp handling.
   // This assert should help detect some regressions early.
-  assert(!FlagSfiDisableCP && "unexpected call to TargetMaterializeConstant");
+  assert((!Subtarget->isTargetNaCl() || FlagSfiEnableCP) && 
+         "unexpected call to TargetMaterializeConstant");
   // @LOCALMOD-END
   EVT VT = TLI.getValueType(C->getType(), true);
 
diff -r 441feeabe2c2 lib/Target/ARM/ARMISelDAGToDAG.cpp
--- a/lib/Target/ARM/ARMISelDAGToDAG.cpp	Mon Mar 28 17:01:56 2011 -0700
+++ b/lib/Target/ARM/ARMISelDAGToDAG.cpp	Tue Jun 28 10:56:00 2011 -0700
@@ -38,7 +38,7 @@
 // @LOCALMOD-START
 #include "llvm/Support/CommandLine.h"
 namespace llvm {
-  extern cl::opt<bool> FlagSfiStore;
+  extern cl::opt<bool> FlagSfiDisableStore;
 }
 // @LOCALMOD-END
 
@@ -420,7 +420,7 @@
   SDValue& N, const ARMSubtarget* Subtarget) {
   assert (N.getOpcode() == ARMISD::Wrapper);
   // Never use this transformation if constant island pools are disallowed 
-  if (FlagSfiDisableCP) return false;
+  if (Subtarget->isTargetNaCl() && (!FlagSfiEnableCP)) return false;
 
   // always apply this when we do not have movt/movw available
   // (if we do have movt/movw we be able to get rid of the
@@ -600,7 +600,7 @@
   // @LOCALMOD-START
   // avoid two reg addressing mode for stores
   const bool is_store = (Op->getOpcode() == ISD::STORE);
-  if (!FlagSfiStore || !is_store ) {
+  if (FlagSfiDisableStore || !is_store ) {
   // @LOCALMOD-END
 
   if (N.getOpcode() == ISD::MUL &&
@@ -683,7 +683,7 @@
   
   // @LOCALMOD-START
   // keep store addressing modes simple
-  if (FlagSfiStore && is_store) {
+  if ((!FlagSfiDisableStore) && is_store) {
     Base = N;
     if (N.getOpcode() == ISD::FrameIndex) {
       int FI = cast<FrameIndexSDNode>(N)->getIndex();
@@ -783,7 +783,7 @@
 
     //if (ConstantSDNode *Sh = dyn_cast<ConstantSDNode>(N.getOperand(1))) {
     ConstantSDNode *Sh = dyn_cast<ConstantSDNode>(N.getOperand(1));
-    if ((!FlagSfiStore || !is_store) && Sh ) { // @LOCALMOD
+    if ((FlagSfiDisableStore || !is_store) && Sh ) { // @LOCALMOD
       ShAmt = Sh->getZExtValue();
       if (isShifterOpProfitable(N, ShOpcVal, ShAmt))
         Offset = N.getOperand(0);
@@ -807,7 +807,7 @@
                                       SDValue &Opc) {
   // @LOCALMOD-START
   const bool is_store = (Op->getOpcode() == ISD::STORE);
-  if (!FlagSfiStore ||!is_store) {
+  if (FlagSfiDisableStore ||!is_store) {
   // @LOCALMOD-END
   if (N.getOpcode() == ISD::SUB) {
 
@@ -851,7 +851,7 @@
   }
 
   // @LOCALMOD-START
-  if (FlagSfiStore && is_store) {
+  if (!FlagSfiDisableStore && is_store) {
     Base = N;
     Offset = CurDAG->getRegister(0, MVT::i32);
     Opc = CurDAG->getTargetConstant(ARM_AM::getAM3Opc(ARM_AM::add, 0),MVT::i32);
@@ -2289,7 +2289,8 @@
                  !ARM_AM::isSOImmTwoPartVal(Val));     // two instrs.
     }
 
-    if (FlagSfiDisableCP) UseCP = false; // @LOCALMOD
+    if (Subtarget->isTargetNaCl() && (!FlagSfiEnableCP)) 
+      UseCP = false; // @LOCALMOD
 
     if (UseCP) {
       SDValue CPIdx =
diff -r 441feeabe2c2 lib/Target/ARM/ARMISelLowering.cpp
--- a/lib/Target/ARM/ARMISelLowering.cpp	Mon Mar 28 17:01:56 2011 -0700
+++ b/lib/Target/ARM/ARMISelLowering.cpp	Tue Jun 28 10:56:00 2011 -0700
@@ -55,8 +55,8 @@
 // @LOCALMOD-START
 #include "llvm/Support/CommandLine.h"
 namespace llvm {
-  extern cl::opt<bool> FlagSfiStore;
-  extern cl::opt<bool> FlagSfiDisableCP;
+  extern cl::opt<bool> FlagSfiDisableStore;
+  extern cl::opt<bool> FlagSfiEnableCP;
 }
 // @LOCALMOD-END
 
@@ -1951,7 +1951,7 @@
   SDValue Chain;
   SDValue Argument;
 
-  if (FlagSfiDisableCP) {
+  if (Subtarget->isTargetNaCl() && (!FlagSfiEnableCP)) {
     // With constant pools "disabled" (moved to rodata), this constant pool
     // entry is no longer in text, and simultaneous PC relativeness
     // and CP Addr relativeness is no longer expressible.
@@ -2043,7 +2043,7 @@
     unsigned ARMPCLabelIndex = AFI->createPICLabelUId();
 
     // @LOCALMOD-BEGIN
-    if (FlagSfiDisableCP) {
+    if (Subtarget->isTargetNaCl() && (!FlagSfiEnableCP)) {
       // Similar to change to LowerToTLSGeneralDynamicModel, and
       // for the same reason.
       unsigned char PCAdj = 0;
@@ -2232,7 +2232,7 @@
   DebugLoc dl = Op.getDebugLoc();
 
   // @LOCALMOD-BEGIN
-  if (FlagSfiDisableCP) {
+  if (Subtarget->isTargetNaCl() && !(FlagSfiEnableCP)) {
     // With constant pools "disabled" (moved to rodata), the constant pool
     // entry is no longer in text, and the PC relativeness is
     // no longer expressible.
@@ -7031,7 +7031,7 @@
 
   // @LOCAMOD-START
   // NOTE: THIS IS A LITTLE DRASTIC
-  if (FlagSfiStore && N->getOpcode() == ISD::STORE) {
+  if (!FlagSfiDisableStore && N->getOpcode() == ISD::STORE) {
     return false;
   }
   // @LOCAMOD-END
@@ -7076,7 +7076,7 @@
     return false;
    // @LOCALMOD-START
   // THIS IS A LITTLE DRASTIC
-  if (FlagSfiStore && N->getOpcode() == ISD::STORE) {
+  if ((!FlagSfiDisableStore) && N->getOpcode() == ISD::STORE) {
     return false;
   }
   // @LOCALMOD-END
diff -r 441feeabe2c2 lib/Target/ARM/ARMNaClHeaders.cpp
--- a/lib/Target/ARM/ARMNaClHeaders.cpp	Mon Mar 28 17:01:56 2011 -0700
+++ b/lib/Target/ARM/ARMNaClHeaders.cpp	Tue Jun 28 10:56:00 2011 -0700
@@ -20,10 +20,10 @@
 
 void EmitSFIHeaders(raw_ostream &O) {
   O << " @ ========================================\n";
-  O << "@ Branch: " << FlagSfiBranch << "\n";
-  O << "@ Stack: " << FlagSfiStack << "\n";
-  O << "@ Store: " << FlagSfiStore << "\n";
-  O << "@ Data: " << FlagSfiData << "\n";
+  O << "@ Branch: " << !FlagSfiDisableBranch << "\n";
+  O << "@ Stack: " << !FlagSfiDisableStack << "\n";
+  O << "@ Store: " << !FlagSfiDisableStore << "\n";
+  O << "@ Data: " << !FlagSfiDisableData << "\n";
 
   O << " @ ========================================\n";
   // NOTE: this macro does bundle alignment as follows
@@ -111,7 +111,7 @@
   }
 
   O << " @ ========================================\n";
-  if (FlagSfiBranch) {
+  if (!FlagSfiDisableBranch) {
     O <<
       "\t.macro sfi_call_preamble cond=\n"
       "\tsfi_nops_to_force_slot3\n"
@@ -143,7 +143,7 @@
 
   }
 
-  if (FlagSfiStore) {
+  if (!FlagSfiDisableStore) {
     O << " @ ========================================\n";
 
     O <<
diff -r 441feeabe2c2 lib/Target/ARM/ARMNaClRewritePass.cpp
--- a/lib/Target/ARM/ARMNaClRewritePass.cpp	Mon Mar 28 17:01:56 2011 -0700
+++ b/lib/Target/ARM/ARMNaClRewritePass.cpp	Tue Jun 28 10:56:00 2011 -0700
@@ -42,16 +42,17 @@
 FlagSfiZeroMask("sfi-zero-mask");
 
 cl::opt<bool>
-FlagSfiData("sfi-data", cl::desc("use illegal at data bundle beginning"));
+FlagSfiDisableData("sfi-data", cl::desc("disable using illegal at data bundle "
+                                        "beginning"));
 
 cl::opt<bool>
-FlagSfiStore("sfi-store", cl::desc("enable sandboxing for stores"));
+FlagSfiDisableStore("sfi-store", cl::desc("disable sandboxing for stores"));
 
 cl::opt<bool> 
-FlagSfiStack("sfi-stack", cl::desc("enable sandboxing for stack changes"));
+FlagSfiDisableStack("sfi-stack", cl::desc("disable sandboxing for stack changes"));
 
 cl::opt<bool>
-FlagSfiBranch("sfi-branch", cl::desc("enable sandboxing for branches"));
+FlagSfiDisableBranch("sfi-branch", cl::desc("disable sandboxing for branches"));
 
 }
 
@@ -630,9 +631,9 @@
        ++MFI) {
     MachineBasicBlock &MBB = *MFI;
 
-    if (FlagSfiStore)  Modified |= SandboxStoresInBlock(MBB);
-    if (FlagSfiBranch) Modified |= SandboxBranchesInBlock(MBB);
-    if (FlagSfiStack)  Modified |= SandboxStackChangesInBlock(MBB);
+    if (!FlagSfiDisableStore)  Modified |= SandboxStoresInBlock(MBB);
+    if (!FlagSfiDisableBranch) Modified |= SandboxBranchesInBlock(MBB);
+    if (!FlagSfiDisableStack)  Modified |= SandboxStackChangesInBlock(MBB);
   }
   DEBUG(LightweightVerify(MF));
   return Modified;
diff -r 441feeabe2c2 lib/Target/ARM/ARMNaClRewritePass.h
--- a/lib/Target/ARM/ARMNaClRewritePass.h	Mon Mar 28 17:01:56 2011 -0700
+++ b/lib/Target/ARM/ARMNaClRewritePass.h	Tue Jun 28 10:56:00 2011 -0700
@@ -16,10 +16,10 @@
 
 namespace llvm {
   extern cl::opt<bool> FlagSfiZeroMask;
-  extern cl::opt<bool> FlagSfiData;
-  extern cl::opt<bool> FlagSfiStore;
-  extern cl::opt<bool> FlagSfiStack;
-  extern cl::opt<bool> FlagSfiBranch;
+  extern cl::opt<bool> FlagSfiDisableData;
+  extern cl::opt<bool> FlagSfiDisableStore;
+  extern cl::opt<bool> FlagSfiDisableStack;
+  extern cl::opt<bool> FlagSfiDisableBranch;
 }
 
 namespace ARM_SFI {
diff -r 441feeabe2c2 lib/Target/ARM/ARMSubtarget.h
--- a/lib/Target/ARM/ARMSubtarget.h	Mon Mar 28 17:01:56 2011 -0700
+++ b/lib/Target/ARM/ARMSubtarget.h	Tue Jun 28 10:56:00 2011 -0700
@@ -23,7 +23,7 @@
 // @LOCALMOD-BEGIN
 #include "llvm/Support/CommandLine.h"
 namespace llvm {
-  extern cl::opt<bool> FlagSfiDisableCP;
+  extern cl::opt<bool> FlagSfiEnableCP;
 }
 // @LOCALMOD-END
 
@@ -224,11 +224,13 @@
 
   bool useMovt() const { return UseMovt && hasV6T2Ops(); }
 
-  // @LOCALMOD
-  bool useConstPool() const { return !FlagSfiDisableCP; }
+  // @LOCALMOD-START
+  bool isTargetNaCl() const {
+    return TargetTriple.getOS() == Triple::NativeClient;
+  }
 
-  // @LOCALMOD
-  bool isTargetNaCl() const { return TargetTriple.getOS() == Triple::NativeClient; }
+  bool useConstPool() const { return (!isTargetNaCl()) || FlagSfiEnableCP; }
+  // @LOCALMOD-END
 
   bool allowsUnalignedMem() const { return AllowsUnalignedMem; }
 
diff -r 441feeabe2c2 lib/Target/ARM/ARMTargetMachine.cpp
--- a/lib/Target/ARM/ARMTargetMachine.cpp	Mon Mar 28 17:01:56 2011 -0700
+++ b/lib/Target/ARM/ARMTargetMachine.cpp	Tue Jun 28 10:56:00 2011 -0700
@@ -26,8 +26,15 @@
 
 // @LOCALMOD-START
 namespace llvm {
-cl::opt<bool> FlagSfiDisableCP("sfi-disable-cp",
-                               cl::desc("disable arm constant island pools"));
+cl::opt<bool> FlagSfiEnableCP("sfi-enable-cp", cl::init(false), cl::Hidden,
+                              cl::desc("enable constant island pools in "
+                                       "text seg for NaCl (NaCl default "
+                                       "for constants is rodata)"));
+
+extern cl::opt<bool> FlagSfiDisableData;
+extern cl::opt<bool> FlagSfiDisableStore;
+extern cl::opt<bool> FlagSfiDisableStack;
+extern cl::opt<bool> FlagSfiDisableBranch;
 }
 // @LOCALMOD-END
 
@@ -193,19 +200,30 @@
     PM.add(createThumb2SizeReductionPass());
 
   // @LOCALMOD-START
-  // Note with FlagSfiDisableCP we effectively disable the
+  // Note with isTargetNaCl() we effectively disable the
   // ARMConstantIslandPass and rely on movt/movw to eliminate the need
   // for constant islands
-  if (FlagSfiDisableCP) {
-    assert(Subtarget.useMovt());
+  if (Subtarget.isTargetNaCl()) {
+    assert(Subtarget.useMovt() && "ARM/NaCl requires movw/movt instruction");
   }
+
+  /// FlagSfi* TRUE implies isTargetNaCl() TRUE
+  assert(((!FlagSfiDisableStore) || Subtarget.isTargetNaCl()) &&
+         "FlagSfiDisableStore can only be true in NaCl");
+  assert(((!FlagSfiDisableData) || Subtarget.isTargetNaCl()) &&
+         "FlagSfiDisableData can only be true in NaCl");
+  assert(((!FlagSfiDisableStack) || Subtarget.isTargetNaCl()) &&
+         "FlagSfiDisableStack can only be true in NaCl");
+  assert(((!FlagSfiDisableBranch) || Subtarget.isTargetNaCl()) &&
+         "FlagSfiDisableStack can only be true in NaCl");
   // @LOCALMOD-END
 
   PM.add(createARMConstantIslandPass());
   
   // @LOCALMOD-START
-  // This pass does all the heavy sfi lifting. 
-  PM.add(createARMNaClRewritePass());
+  // This pass does all the heavy sfi lifting for ARM/SFI
+  if (Subtarget.isTargetNaCl())
+    PM.add(createARMNaClRewritePass());
   // @LOCALMOD-END
  
   return true;
diff -r 441feeabe2c2 lib/Target/ARM/ARMTargetMachine.h
--- a/lib/Target/ARM/ARMTargetMachine.h	Mon Mar 28 17:01:56 2011 -0700
+++ b/lib/Target/ARM/ARMTargetMachine.h	Tue Jun 28 10:56:00 2011 -0700
@@ -32,7 +32,7 @@
 // @LOCALMOD-START
 #include "llvm/Support/CommandLine.h"
 namespace llvm {
-   extern cl::opt<bool> FlagSfiDisableCP;
+   extern cl::opt<bool> FlagSfiEnableCP;
 }
 // @LOCALMOD-END
 
